[e2e] Refactor discussions

tuliobluz opened the pull request:
## PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)
- [ ] - Unit and e2e tests for the changes that have been added (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix (fixes an issue)
- [ ] Feature (adds functionality)
- [ ] Code style update
- [x] Refactoring (no functional changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation changes
- [ ] Other... Please describe:

## What is the current behavior?

- There are a lot of duplication codes in discussion scenarios that could be refactored in the `commandsUi.ts`
- Missing assertion to deleteComments amount

## What is the new behavior?

- Create addLastComment, editLast, deleteLastCommentOrReply, addReplytoFirstComment to avoid duplicate code
- Add `cy.get('[data-cy=edit-comment]').should('not.exist')` to cypress does not take the text from the edit modal, making sure the test checks the new comment in the right place, instead of the modal.
- Add a last assertion to validate that deleted comments are absent in the total amount.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

benfurber commented on packages/cypress/src/support/commandsUi.ts:
```diff
@@ -34,7 +34,12 @@ interface IOpeningTime {
+      addLastComment(newComment: string): Chainable<void>
+      addReplytoFirstComment(reply: string): Chainable<void>
       clickMenuItem(menuItem: UserMenuItem): Chainable<void>
+      checkCommentItem(firstComment: string, length: number): Chainable<void>
```
`firstComment` -> `comment`

benfurber commented on packages/cypress/src/integration/howto/discussions.spec.ts:
```diff
@@ -17,50 +17,39 @@ describe('[Howto.Discussions]', () => {
   it('allows authenticated users to contribute to discussions', () => {
     const newComment = 'An interesting howto. The answer must be...'
     const updatedNewComment =
       'An interesting howto. The answer must be that when the sky is red, the apocalypse _might_ be on the way.'
-    const newReply = 'Thanks Dave and Ben. What does everyone else think?'
-    const updatedNewReply = 'Anyone else?'
+    const reply = 'Thanks Dave and Ben. What does everyone else think?'
+    const updatedReply = 'Anyone else?'
```
Either comment and reply should be prefixed 'new' or not. Ending up `newComment` and `reply` is my least favourite end case. 

My preference would be to keep the new prefix. My intent was to highlight that comments were already present before the spec adds them. Plus not changing `newReply` to `reply` keeps the diff smaller.

benfurber commented on packages/cypress/src/integration/howto/discussions.spec.ts:
```diff
@@ -17,50 +17,39 @@ describe('[Howto.Discussions]', () => {
-      .scrollIntoView()
-      .should('be.inViewport', 10)
-    cy.contains(firstComment.text)
+    cy.checkCommentItem(firstComment.text, 2)
```
I don't think I like this command as is...

* It's checking two things, the scroll view and the @ mention comment content.
* It's more an expectation than a ui command, so it should probably be in commands rather than commands ui.

So let's go with two commands in the other file.

benfurber reviewed (changes requested):
Nice one @tuliobluz. 

My main feedback is around the naming. Rather than specifying last or first I had two thoughts first was having last as a default value like `selectTag` does. But then I also thought as we're not using that method in a different context, that's building functionality we won't use yet. So if you want to do that, go for it, otherwise suggested names below:

* `addLastComment` -> `addComment`
* `addReplytoFirstComment` -> `addReply`
* `deleteLastCommentOrReply` -> `deleteDiscussionItem`
* `editLast` -> `editDiscussionItem`

Another thought. I think all the commands should start with the cy step included (generally that's the pattern we've gone with).

The doc queries for latestCommentDate and notifications could deffo be commands too.

Oh and obvs my comments on the howto spec apply to the other two.

tuliobluz commented:
@benfurber Addressed all the comments, please review them again. I am not being able to squash the commits 

When I am trying I get this error message

`refactor: yarn format
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]
⚠   body must have leading blank line [body-leading-blank]
`

benfurber commented on packages/cypress/src/integration/research/discussions.spec.ts:
```diff
@@ -71,17 +58,20 @@ describe('[Research.Discussions]', () => {
+    cy.contains(updatedNewReply).should('not.exist')
     cy.contains(`1 Comment`)
+
+    cy.step('Check comments number after deletation')
```
Typo: deletion

benfurber reviewed (changes requested):
Nice.

I think the new commands are still missing a `cy.step()`?

What command with the squash issue were you running? I find using fixup in an interactive rebase works best.

tuliobluz commented:
I did not add the steps because the test files have `cy.step` then it is easy to read the file, however, I can move to there `commandsUi`

benfurber commented:
@tuliobluz The error message suggests it's more a problem with the commit message than the squashing.
